src/link-preview-expander.ts

```diff
@@ -1,16 +1,194 @@
+                        metadataCache.set(maybeURL, metadata)
+                        return createResult(metadata)
+                    })
+                    .catch(() => {
+                        metadataCache.reportFailure(maybeURL)
+                        return createResult()
+                    })
+            },
         })
     )
 }
 
+export const metadataAttributes = ['image', 'title', 'description'] as const
+type MetadataAttributes = typeof metadataAttributes[number]
+
+export type Metadata = Record<MetadataAttributes, string>
+
+interface MetadataProvider<T = string> {
+    type: T
+    selectorType: 'property' | 'name'
+    selectorPrefix: string
+}
+
+type MetadataProviderType = 'openGraph' | 'twitter' | 'default'
+
+export type MetadataByProvider = Record<MetadataProviderType, Metadata>
+
+// In order of priority in the 'metadata cascade'
+export const metadataProviders: MetadataProvider<MetadataProviderType>[] = [
+    {
+        type: 'openGraph',
+        selectorType: 'property',
+        selectorPrefix: 'og:',
+    },
+    {
+        type: 'twitter',
+        selectorType: 'name',
+        selectorPrefix: 'twitter:',
+    },
+    {
+        type: 'default',
+        selectorType: 'name',
+        selectorPrefix: '',
+    },
+]
+
+/**
+ * Merges metadata by priority of provider type.
+ *
+ * @param metadataByProvider Record of Metadata objects keyed by metadata provider type
+ */
+export function mergeMetadataProviders(metadataByProvider: MetadataByProvider): Metadata {
+    const finalMetadata = initializeMetadata()
+
+    for (const attribute of metadataAttributes) {
+        for (const { type } of metadataProviders) {
+            const value = metadataByProvider[type][attribute]
+            if (value) {
+                finalMetadata[attribute] = value
+                break
+            }
+        }
+    }
+
+    return finalMetadata
+}
+
+/**
+ * Retrieves metadata from all providers given an HTML string
+ */
+export function getMetadataFromHTMLString(htmlString: string): MetadataByProvider {
+    const root = parse(htmlString)
+
+    const metadataByProvider: MetadataByProvider = {
+        openGraph: initializeMetadata(),
+        twitter: initializeMetadata(),
+        default: initializeMetadata(),
+    }
+
+    if (!root.valid) {
+        return metadataByProvider
+    }
+
+    // node-html-parser doesn't support this selector: 'meta[property="og:image"]'
+    const metaElements = root.querySelectorAll('meta')
+
+    outer: for (const metaElement of metaElements) {
+        for (const { type, selectorType, selectorPrefix } of metadataProviders) {
+            for (const attribute of metadataAttributes) {
+                const propertyOrName = metaElement.getAttribute(selectorType)
+                if (propertyOrName === selectorPrefix + attribute) {
+                    metadataByProvider[type][attribute] = metaElement.getAttribute('content') ?? ''
```

Comment:
We first parse all the meta attributes into this interface structure, to then call the merge function that then only uses one of the providers. Would it be simpler to just read the meta tags from the HTML directly in priority order and return the first valid one? And have the providers just be hardcoded instead of in an object structure?

Reply:
How would we know if a higher priority provider's tag comes later in the array? That was my thought process in writing this, but it's useless; I think Twitter and Open Graph tags have the same content most of the time, so there doesn't need to be any concept of priority. I think we should keep the object structure as it would be simpler without the intermediate object (simply `{ selectorType: 'property' | 'name'; selectorPrefix: string }`.